enhancements/machine-config/on-cluster-builds.md

```diff
@@ -0,0 +1,1083 @@
+  - joelspeed
+creation-date: 2023-11-06
+last-updated: 2023-11-06
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/MCO-834
+see-also:
+  - "/enhancements/ocp-coreos-layering.md"
+replaces:
+  - ""
+superseded-by:
+  - ""
+---
+
+# On-Cluster Builds
+
+## Summary
+
+Change the Machine Config Operator (MCO) that is shipped in OpenShift to
+perform an on-cluster build (OCB) to combine the base OS image shipped
+with OpenShift, any admin-provided custom content, and the MachineConfig
+content managed by the MCO. This image will then be deployed to all of
+the cluster nodes using rpm-ostree (and later, bootc) to facilitate a
+singular, transactional, update.
+
+With image-based RHEL becoming a reality, on-cluster builds will need to evolve
+to accommodate this new paradigm.
+
+## Motivation
+
+### User Stories
+
+#### Device Drivers For Specialty Hardware
+
+As an OpenShift cluster administrator, I want to install a device driver,
+ideally on a Day 0 basis, to ensure that my cluster can use its underlying
+hardware to the fullest potential right from installation time.
+
+#### Software Agents To Ensure Compliance
+
+As an OpenShift cluster administrator, I want to install an intrusion
+detection system (IDS) agent so that workloads on my cluster are in
+compliance with regulatory and oversight requirements for my industry.
+
+#### Keep Custom OS Content Up-To-Date
+
+As an OpenShift cluster administrator, I want my customized OS content
+to be kept up-to-date with my current and future OpenShift releases
+automatically.
+
+#### More Control Over Config Change Deployments
+
+As an OpenShift cluster administrator, I want to prepare my customized
+OS content and configuration changes ahead of time so that I can reduce
+or eliminate issues with deploying them.
+
+#### Managed OpenShift Service Provider
+
+As a Managed OpenShift SRE, I want to install cloud specific tooling to ensure
+SRE can triage issues with all nodes involved in the lifecycle of a Managed
+OpenShift cluster, including bootstrap, without requiring any reboot/pivot.
+
+#### Managed OpenShift Service Consumer
+
+As a Managed OpenShift service consumer, I want to know any changes to node
+configuration can be tested and validated against a known set of criteria so I
+know my nodes will not break support agreements with Managed OpenShift Service
+Providers.
+
+### Goals
+
+- A cluster administrator will be able to install and upgrade any
+  additional custom OS content required for their clusters’ operating
+  system.
+
+- Any custom OS content will be kept up-to-date throughout the lifecycle
+  of a given OpenShift cluster. In other words, whenever the cluster is
+  upgraded to a new OpenShift release, any custom OS content will be
+  upgraded as well (subject to update availability).
```

Comment:
What is custom OS content here? How can we automatically update it?

Reply:
Yes, I think this is a key point. How and when we rebuild is a very important thing for nontrivial workloads. Consider things like 3rd party agents from 3rd party yum repos.